hack/tools/plutonodashboardcheck/main.go

```diff
@@ -0,0 +1,231 @@
+// Copyright 2023 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
```

Comment:
Is it described somewhere why 
1. we don't use https://github.com/grafana/dashboard-linter
2. why we don't check only for json validity with a tool like `jq`

and why exactly we need such go program that checks for json validity and for additional things?

When creating https://github.com/gardener/gardener/issues/7761 I didn't mean to introduce a go binary on our side that we have to maintain to perform such checks.
I would rather prefer keeping the things simple by reusing https://github.com/grafana/dashboard-linter (if possible) or checking for JSON validity only with already existing tool like jq.

I would like to hear also what @istvanballok and @rickardsjp you think about it.

Reply:
I agree that I'd also prefer a simpler approach.

I think checking the JSON syntax with jq and even the existence of a title field and a globally unique uid is feasible.

What do you think about this approach with a shell script with jq and awk?

```bash
set -o pipefail

for i in $(find . -path '*/seed-bootstrap/dashboards/*' -name '*.json')
do 
  jq -r '{title: (.title // error("title is not set")),
          uid:   (.uid   // error("uid is not set"))}' "$i" \
  || { echo "Failure in dashboard $i" >&2; return 1; }
done \
| jq '.uid' | sort | uniq -c \
| awk '$1 > 1 {duplicatefound=1
               printf "%s is a duplicate dashboard UID\n", $0}
       END {exit duplicatefound}'
```

It could run similarly for the `seed-monitoring` path.

It asserts that the json file can be parsed by jq, it has a title and a uid field, and that the uid is unique. It returns a non zero exit code if one of the failures occur and produces a log message that hints at the problem.

Possible failure outcomes:

```
parse error: Invalid numeric literal at line 1, column 2
Failure in dashboard ./charts/seed-bootstrap/dashboards/vpa-recommender.json

jq: error (at ./charts/seed-bootstrap/dashboards/vpa-recommender.json:1195): uid is not set
Failure in dashboard ./charts/seed-bootstrap/dashboards/vpa-recommender.json

2 "vpa" is a duplicate dashboard UID
```

Reply:
A slight variation of my snippet above:

- avoid https://www.shellcheck.net/wiki/SC2044
- check all the dashboard files (both seed-monitoring and seed-bootstrap), but skip the symbolic links
- improve the error message of the duplicate UID-s (2 dashboards that were copied for the workerless case would need to be adjusted, although they are not loaded into Plutono at the same time, but we could enforce the convention that all the dashboards should have a unique uid)

```bash
set -o pipefail

function check_dashboards {
  while IFS= read -r -d '' file
  do 
    jq -c -r '{title: (.title // error("title is not set")),
               uid:   (.uid   // error("uid is not set"))}
              | "\(input_filename) \(.uid)"' "$file" \
    || { echo "Error: Failure while parsing dashboard: $file" >&2; return 1; }
  done < <(find . -path '*/dashboards/*' -name '*.json' -type f -print0) \
  | sort -k 2 | uniq -D -f 1 \
  | { grep . && echo "Error: dashboards with duplicate UIDs" >&2 && return 1 || return 0; }
}

check_dashboards
```

Possible failure messages:

```text
parse error: Invalid numeric literal at line 1, column 2
Error: Failure while parsing dashboard ./charts/seed-bootstrap/dashboards/vpa-recommender.json

jq: error (at ./charts/seed-bootstrap/dashboards/vpa-recommender.json:1195): uid is not set
Error: Failure while parsing dashboard: ./charts/seed-bootstrap/dashboards/vpa-recommender.json

./charts/seed-monitoring/charts/plutono/dashboards/owners/worker/cluster-overview-dashboard.json cluster-overview
./charts/seed-monitoring/charts/plutono/dashboards/owners/workerless/cluster-overview-dashboard.json cluster-overview
./charts/seed-monitoring/charts/plutono/dashboards/owners/worker/kubernetes-control-plane-status-dashboard.json kubernetes-control-plane-status
./charts/seed-monitoring/charts/plutono/dashboards/owners/workerless/kubernetes-control-plane-status-dashboard.json kubernetes-control-plane-status
Error: dashboards with duplicate UIDs
```

Reply:
Example output for the original motivation of this issue:

https://github.com/gardener/gardener/issues/7747#issuecomment-1497391411

```text
$ check_dashboards
parse error: Expected separator between values at line 1450, column 15
Error: Failure while parsing dashboard: ./charts/seed-bootstrap/dashboards/istio/istio-ingress-gateway-dashboard.json
```

Reply:
> why we don't use https://github.com/grafana/dashboard-linter

Yeah, at the beginning i looked through the `dashboard-linter` and thinking to use it. It implements multiple rules which we would not use and not have the basic check like empty title or unique ID.  So i changed the plan to own go package.